src/script/interpreter.cpp

```diff
@@ -1929,6 +1937,178 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                 }
                 break;
 
+                case OP_ADD64:
+                case OP_SUB64:
+                case OP_MUL64:
+                case OP_DIV64:
+                case OP_LESSTHAN64:
+                case OP_LESSTHANOREQUAL64:
+                case OP_GREATERTHAN64:
+                case OP_GREATERTHANOREQUAL64:
+                case OP_EQUAL64:
+                case OP_AND64:
+                case OP_OR64:
+                case OP_XOR64:
+                {
+                    // Opcodes only available post tapscript
+                    if (sigversion == SigVersion::BASE || sigversion == SigVersion::WITNESS_V0) return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
+
+                    if (stack.size() < 2)
+                        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+
+                    valtype& vchb = stacktop(-1);
+                    valtype& vcha = stacktop(-2);
+                    if (vchb.size() != 8 || vcha.size() != 8)
+                        return set_error(serror, SCRIPT_ERR_EXPECTED_8BYTES);
+
+                    int64_t b = ReadLE64(vchb.data());
+                    int64_t a = ReadLE64(vcha.data());
+
+                    switch(opcode)
+                    {
+                        case OP_SUB64:
+                            b = -b;
+                            // fallthrough (a - b == a + (-b)). Negation should not change any overflow logic
+                        case OP_ADD64:
+                            if ((a > 0 && b > std::numeric_limits<int64_t>::max() - a) ||
+                                (a < 0 && b < std::numeric_limits<int64_t>::min() - a))
```

Comment:
In theory these limits are implementation defined.

I would prefer you use the constants `0xFFFFFFFFFFFFFFFF` and `-0x10000000000000000`.

cc @roconnor-blockstream what do you think?

Reply:
These constants are also used in bench tests. AWaiting response from Russell before changing.

Reply:
How are they implementation defined?  Are you worried about non-two's complement?

Reply:
I am worried about non-two's-complement, yes. And I say "implementation defined" because I could not find a citation saying that these constants should have specific values.

I'd appreciate a static assert that these values are what we expect, but I don't feel too strongly about this. In practice I doubt that our software (or Bitcoin) will work properly on a twos-complement system, and I'm not aware of any real-world users with such systtems.

Reply:
Static_asserts are still needed.

Reply:
Added at the top of the file